File: go/vector.go
package main

import "math"

type Point = [3]float64
type Vec = [3]float64
type Color = [3]float64

type Ray struct {
    Start Point
    Dir Vec
}

func VNeg(v Vec) Vec {
    var ret Vec
    for i := 0; i < 3; i++ {
        ret[i] = -v[i]
    }
    return ret
}

func VAdd(v, w Vec) Vec {
    var ret Vec
    for i := 0; i < 3; i++ {
        ret[i] = v[i] + w[i]
    }
    return ret
}

func VSub(v, w Vec) Vec {
    var ret Vec
    for i := 0; i < 3; i++ {
        ret[i] = v[i] - w[i]
    }
    return ret
}

func VMul(c float64, v Vec) Vec {
    var ret Vec
    for i := 0; i < 3; i++ {
        ret[i] = c * v[i];
    }
    return ret
}

func DotProduct(v, w Vec) float64 {
    total := 0.0
    for i := 0; i < 3; i++ {
        total += v[i] * w[i]
    }
    return total
}

func Magnitude(v Vec) float64 {
    return math.Sqrt(DotProduct(v, v))
}

func Normalize(v Vec) Vec {
    return VMul(1.0 / Magnitude(v), v)
}

func Project(v, w Vec) Vec {
    return VMul(DotProduct(v, w) / DotProduct(w, w), w)
}

File: nim/src/vector.nim
import std/math

type
    Vector* = array[0..2, float]
    Point* = Vector
    Color* = Vector

func `+`*(u, v: Vector): Vector =
    for i in 0..2: result[i] = u[i] + v[i]

proc `+=`*(u: var Vector, v: Vector) =
    for i in 0..2: u[i] += v[i]

func `-`*(u, v: Vector): Vector =
    for i in 0..2: result[i] = u[i] - v[i]

func `-`*(v: Vector): Vector =
    for i in 0..2: result[i] = -v[i]

func `*`*(x: float, v: Vector): Vector =
    for i in 0..2: result[i] = x * v[i]

func dot*(u, v: Vector): float =
    for i in 0..2: result += u[i] * v[i]

func magnitude*(v: Vector): float =
    return sqrt(dot(v, v))

func normalize*(v: Vector): Vector =
    return (1.0 / v.magnitude) * v

func project*(u, v: Vector): Vector =
    return dot(u, v) / dot(v, v) * v
